src/scf/potential_mixing.jl

```diff
@@ -0,0 +1,345 @@
+
+    # Accept model if it leads to minimum and is either tight or shows a negative slope
+    if minimum_exists && (tight_model || (slope < -eps(T) && trusted_model))
+        α_model = -slope / curv
+        mpi_master() && @debug "Quadratic model accepted" model_relerror slope curv α_model
+        return α_model
+    else
+        mpi_master() && @debug "Quadratic model discarded" model_relerror slope curv
+        nothing  # Model not trustworthy ... better return nothing
+    end
+end
+
+# Adaptive damping using a quadratic model
+@kwdef struct AdaptiveDamping
+    α_init = 0.5       # The initial damping value (used as first trial value in first SCF step)
+    α_min = 0.01       # Minimal damping
+    α_max = 1.0        # Maximal damping
+    α_trial_min = 0.1  # Minimal first trial damping used in a step
+    α_trial_enhancement = 1.1  # Enhancement factor to α_trial in case a step is immediately successful
+    modeltol = 0.1     # Maximum relative error for model to be considered trustworthy
+end
+
+function propose_backtrack_damping(damping::AdaptiveDamping, info, info_next)
+    if abs(info_next.α) < 1.75damping.α_min
+        # Too close to α_min to be worth making another step ... just give up
+        return info_next.α
+    end
+
+    α = scf_quadratic_model(info, info_next; modeltol=damping.modeltol)
+    isnothing(α) && (α = info_next.α / 2)  # Model failed ... use heuristics
+
+    # Adjust α to stay within desired range
+    α_sign = sign(α)
+    α = clamp(abs(α), damping.α_min, damping.α_max)
+    α = min(0.95abs(info_next.α), α)  # Avoid to get stuck
+    return α_sign * α
+end
+
+initial_damping(damping::AdaptiveDamping) = damping.α_init
+
+function next_trial_damping(damping::AdaptiveDamping, info, info_next, step_successful)
+    n_backtrack = length(info_next.diagonalization)
+
+    α_trial = abs(info_next.α)  # By default use the α that worked in this step
+    if step_successful && n_backtrack == 1  # First step was good => speed things up
+        α_trial ≥ damping.α_max && return damping.α_max  # No need to compute model
+        α_model = scf_quadratic_model(info, info_next; modeltol=damping.modeltol)
+        if !isnothing(α_model)  # Model is meaningful
+            α_trial = max(damping.α_trial_enhancement * abs(α_model), α_trial)
+        end
+    end
+
+    # Ensure returned damping is in valid range
+    clamp(α_trial, damping.α_trial_min, damping.α_max)
+end
+
+struct FixedDamping
+    α
+end
+FixedDamping() = FixedDamping(0.8)
+propose_backtrack_damping(damping::FixedDamping) = damping.α
+initial_damping(damping::FixedDamping) = damping.α
+next_trial_damping(damping::FixedDamping, info, info_next, successful) = damping.α
+
+
+# Notice: For adaptive damping to run smoothly, multiple defaults need to be changed.
+#         See the function scf_potential_mixing_adaptive for that use case.
+@timing function scf_potential_mixing(
```

Comment:
very annoying that we have to duplicate all these things. We should revisit the SCF & diagonalization parameter structs at some point

Reply:
Agree. I think what we will need is also a generic "do scf" routine that dispatches to the various implementations we have (newton, potential mixing, density mixing etc). I was planning of writing an overview page contrasting the implementations as well. It starts to get a bit messy.